### PR TITLE
DL: Add support for custom loss functions

### PR DESCRIPTION
**Pull request by khannaekta**
JIRA: https://issues.apache.org/jira/browse/MADLIB-1441

For MADlib deep_learning module to support custom functions, we have the following functions are affected:
1. madlib_keras_fit()
2. madlib_keras_fit_multiple_model()
3. madlib_keras_evaluate()
Currently the compile_param passed in are like
 `optimizer=SGD(lr=0.01, decay=1e-6, nesterov=True), loss='categorical_crossentropy', metrics=['mae']`
For a custom function we would just pass in as follows:
 `optimizer=SGD(lr=0.01, decay=1e-6, nesterov=True), loss=custom_fn1, metrics=['mae']` 

Object table definition with a dill object of custom functions is as follows:
|id|name|description|function|
|---|---|---|---|
|1|custom_fn1|test function|\x80026...|
|2|custom_fn2|test function2|\x8ab26...|

For fit_multiple, we want to support multiple configs in mst_table with a mix of custom and keras loss functions:
|mst_key|model_id|compile_params|fit_params|
|---|---|---|---|
|2|1|loss='custom_fn1', optimizer=....|batch_size=16, epochs=1|
|3|1|loss='custom_fn2', optimizer=....|batch_size=16, epochs=1|
|1|1|loss='categorical_crossentropy', optimizer=....|batch_size=16, epochs=1|

Following are the relevant API Changes
 **Fit**: add a new optional param for passing in the object_table name
 object_table (optional) VARCHAR: Name of the table containing Python objects in the case that custom loss functions or custom metrics are specified in the parameter `compile_params
```
madlib_keras_fit(
    source_table,
    model,
    model_arch_table,
    model_id,
    compile_params,
    fit_params,
    num_iterations,
    use_gpus,
    validation_table,
    metrics_compute_frequency,
    warm_start,
    name,
    description,
    object_table  -- new parameter
    )
```
**Fit_multiple**: No change to the `madlib_keras_fit_multiple_model()` function. Reads object_table information from the model_selection table.
 A summary table named <model>_summary is also created, which has the following new columns:
|model_selection_table |Name of the table containing model selection parameters to be tried.|
|object_table| Name of the object table containing the serialized Python objects for custom loss functions and custom metrics (read from the mst_summary table).|

**Evaluate**: No change to the `madlib_keras_evaluate()` function. Reads object_table information from the model table. Output table adds a new column
|loss_type|Type of loss used that was used in the training step
 If a custom loss or metric is used, we should give the name of it. Otherwise list the built-in one used|

- [ ] Add the module name, JIRA# to PR/commit and description.
- [ ] Add tests for the change.

### CODE REVIEW

File: src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in [hunk 1]
```diff
@@ -175,6 +179,17 @@ def get_metrics_from_compile_param(str_of_args):
                         "please refer to the documentation").format(ckey))
     return metrics
 
+def get_loss_from_compile_param(str_of_args):
+    compile_dict = convert_string_of_args_to_dict(str_of_args)
+    loss = None
+    ckey = 'loss'
```
Comment (orhankislal) [inline src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in:185]: We use `'loss'` in a number of different places and this function is literally named for loss. I don't think we need the `ckey` variable. If we want to generalize this function as in `get_key_from_compile_param(str_args,key)` that might work as well.

File: src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in [hunk 2]
```diff
@@ -33,6 +33,8 @@ import keras.losses as losses
 
 import madlib_keras_serializer
 import madlib_keras_gpu_info
+from madlib_keras_custom_function import CustomFunctionSchema
```
[the remark this answers is not in the document]
Comment (khannaekta) [reply src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in:36]: Thanks for mentioning. The plan is to squash all the commits together and merge it as 1 commit, kept them separate for the easy PR review.

File: src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in [hunk 3]
```diff
@@ -331,3 +337,32 @@ def validate_compile_param_types(compile_dict):
             compile_dict['sample_weight_mode'] is None or
             compile_dict['sample_weight_mode'] == "temporal",
             """compile parameter sample_weight_mode can only be "temporal" or None""")
+
+# Returns an object of custom function name and it corresponding object
+def query_custom_functions_map(object_table, custom_fn_names):
+    if len(custom_fn_names) < 1:
```
[the remark this answers is not in the document]
Comment (khannaekta) [reply src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in:370]: The `custom_fn_names` are passed in as what is returned from `get_custom_functions_list()` which is returns a list (empty or with values). Therefore, skipped checking for`None` for now.

**Comment by fmcquillan99** [comment]
Tests look good so far.  Once we get the custom metrics feature in, I will do more extensive testing and update the user docs.

LGTM
